### PR TITLE
feat: add vault_replace_span and vault_insert_at_anchor tools

### PR DESCRIPTION
**Pull request by aliasunder**
## Summary

- **`vault_replace_span`** — replace a contiguous block of lines identified by short anchor substrings with new content, in one atomic write. Same anchor semantics as `vault_delete_span` (case-sensitive substring, ambiguity-is-error, `first_match` escape hatch), with `collapseBlankRuns` on the reassembled body. Replaces the fragile two-step `delete_span` + `patch_note` workaround.
- **`vault_insert_at_anchor`** — insert content before or after a line identified by anchor, freeing edits from heading-only targeting. No blank-line collapse (insertion doesn't create gaps).
- Together with `vault_delete_span`, every line-level edit (delete, replace, insert) can now be targeted by a short anchor substring instead of exact text, addressing ~82/month edit-side guess-miss errors (49 `replace_in_note` text-not-found + 33 `delete_span` anchor-not-found from the Jul 9 – Aug 6 usage window).

## Design decisions

- **Two new tools, purely additive** — no existing tool semantics change. `vault_replace_in_note` (inline edits), `vault_patch_note` (heading-targeted), and `vault_delete_span` (anchor deletion) keep their distinct roles.
- **`vault_insert_at_anchor`** over `vault_insert_span` — descriptive accuracy over family-suffix consistency; "span" doesn't apply to an insert.
- **Non-empty content required** (`.min(1)`) on both tools — `vault_delete_span` exists for deletion.
- **`collapseBlankRuns` on `vault_replace_span`** (same as `delete_span`) — prevents 3+ consecutive blank-line runs at the seam. Not applied on `vault_insert_at_anchor` (insertion doesn't create gaps).
- **`vault_insert_at_anchor` is annotated as an additive write** (`destructiveHint: false`) — it can only add lines, never overwrite or remove them, matching `vault_create_task`. `vault_replace_span` stays destructive. The shared `ADDITIVE_WRITE_ANNOTATIONS` constant now backs both additive tools.
- **One span resolver for the family** — `resolveSpanLines` resolves `{ startLine, endLine }` for `deleteSpan` and `replaceSpan`; all three ops splice with `Array.prototype.toSpliced`.
- **Cross-references gated** — every mention of a sibling tool in the three anchor-tool descriptions goes through `whenToolEnabledText`, so a reference disappears when its target is disabled.
- **Descriptions state the non-obvious semantics** — first-edit YAML normalization (same caveat as the sibling edit tools), trailing-newline behaviour of `content` (adds a blank line), verbatim insert vs. collapsed replace, what each `position` value does, and the literal confirmation-message shape each tool returns.

## Changes

| File | Change |
| --- | --- |
| `vault-patcher.ts` | `replaceSpan` + `insertAtAnchor`; shared `resolveSpanLines`; `toSpliced` in all span ops |
| `tool-registry.ts` | 2 new `TOOL_NAMES` + `TOOL_REGISTRY` entries; `ADDITIVE_WRITE_ANNOTATIONS` |
| `vault-crud-tools.ts` | 2 new handler registrations + cross-reference and parameter-semantics updates; `vault_patch_note` states its no-separator `content` semantics and literal return shape |
| `vault-patcher.test.ts` | 30 new unit tests (happy paths, anchors, frontmatter exclusion, edge cases, errors, control characters, concurrency) |
| `tool-registry.test.ts` | Literal annotation spot-checks for both new tools |
| `server-integration.test.ts` | Tool counts 31→33 (assertions and section labels), write chain extended, read-only exclusions |
| `server-error-contracts.test.ts` | 12 new error contract tests (traversal, hidden, not-found, anchor, extension) |
| `README.md` | 2 new rows in tools table |
| `ARCHITECTURE.md` | 2 new rows; edit tools listed by targeting mode (heading / exact text / anchor) |
| `AGENTS.md` | Structure tree updated (9→11 tools) |
| `DOCKERHUB.md` | Regenerated |

## Test plan

- [x] 3,287 tests pass (83 files) — includes the new unit, integration, error-contract, and registry tests
- [x] `npm run build` clean (server + CLI)
- [x] `npm run lint` — 0 errors
- [x] Emitted `tools/list` JSON Schema inspected from a booted server for both tools — `required` sets, string `enum` for `position`, `minLength` floors, parameter descriptions, and annotations all as intended
- [x] CI checks (arch-smoke, lint, test, integration)
- [x] Live validation on a deployed `:remote` build of this branch (`6e2a2d4`; the two later commits are docs-only), driven through an MCP client against a real vault — 26 calls to the two new tools on a throwaway note, every result read back and compared against the documented contract:
  - `vault_replace_span` writes: single-line table row; multi-line callout with `end_anchor`; `start_anchor` and `end_anchor` on the same line; trailing newline in `content` adds one blank line and the seam collapses to a single blank; leading/trailing blank padding in `content` collapses; `first_match` takes the first of two matches; frontmatter preserved on every write; a property value matching the anchor text is not a match (body only).
  - `vault_replace_span` error paths (exact messages): start anchor not found; ambiguous start anchor (2 lines); ambiguous end anchor at or after the start; end anchor only above the start → "end anchor not found … at or after the start anchor"; note not found; hidden path blocked; anchor present only in frontmatter → not found; path without `.md`.
  - `vault_insert_at_anchor` writes: `after` a table row; `before` a heading with a trailing newline (callout + blank line); `before` with `first_match` on a duplicated list item; `after` the last line with a multi-line block whose internal blank line is kept verbatim and no collapse at the seam.
  - `vault_insert_at_anchor` error paths: anchor not found; ambiguous anchor; note not found; hidden path blocked; anchor present only in frontmatter → not found; path without `.md`; invalid `position` rejected at the schema layer (`-32602`).
  - No regressions in the surrounding surface on the same note: `vault_delete_span`, `vault_replace_in_note`, `vault_patch_note`, `vault_create_task`, `vault_update_task`, `vault_read_note` (full + heading), plus `vault_search`, `vault_list_tasks`, `vault_get_backlinks`, `vault_get_daily_note`, `vault_memory_recall`, `vault_list_files` — all behaved as before.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added anchor-based editing tools to replace content between anchors or insert content before or after an anchored line.
  - Preserves document frontmatter and supports multi-line content with validation and safe error handling.

- **Documentation**
  - Updated tool listings and architecture documentation to describe the new editing capabilities.

- **Tests**
  - Expanded coverage for successful edits, validation, missing anchors/files, hidden paths, ambiguous anchors, and concurrent writes.

### CODE REVIEW

File: src/vault-mcp/vault-operations/vault-patcher.ts [hunk 1]
```diff
@@ -489,9 +489,157 @@ const deleteSpan = async (
...
+          })
+
+    const replacedLines = lines.slice(startLine, endLine + 1)
+    const contentLines = content.split("\n")
+    const updatedLines = [
+      ...lines.slice(0, startLine),
+      ...contentLines,
+      ...lines.slice(endLine + 1),
+    ]
+    const normalizedBody = collapseBlankRuns(updatedLines.join("\n"))
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/vault-patcher.ts:560]: This is intentional — `collapseBlankRuns` on the full body is a deliberate design decision made during planning. The user was asked explicitly whether to collapse at the seam or not, and chose "Same as delete_span" (apply `collapseBlankRuns` on the full reassembled body). The tradeoff (content containing 3+ consecutive blank lines gets collapsed) was presented and accepted — no Obsidian content legitimately uses 3+ consecutive blank lines.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/mcp-core/tools/vault-crud-tools.ts [hunk 2]
```diff
@@ -558,7 +558,7 @@ Returns: Confirmation message. A no-heading prepend that nested existing content
 Example: vault_replace_in_note({ path: "Projects/plan.md", old_text: "TODO: write summary", new_text: "Summary complete." })
 
 When to use: Targeted text changes within a single location — fixing typos, updating values, renaming terms, or removing a short line (new_text=""). Replaces text in place; does not move content across sections.
-To delete a large multi-line block, prefer vault_delete_span (short anchors instead of full old_text). To relocate content between headings, vault_patch_note to add at the target first, then remove from source (new_text="") — add-before-delete, so a failure duplicates the block instead of losing it.
+To delete a large multi-line block, prefer vault_delete_span (short anchors instead of full old_text).${whenToolEnabledText("vault_replace_span", " To replace a large block by anchors instead of reproducing the full old_text, use vault_replace_span.")} To relocate content between headings, vault_patch_note to add at the target first, then remove from source (new_text="") — add-before-delete, so a failure duplicates the block instead of losing it.
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/mcp-core/tools/vault-crud-tools.ts:591]: This is intentional — when `vault_replace_span` is disabled, the `whenToolEnabledText` sentence disappears and the text reads: "To delete a large multi-line block, prefer vault_delete_span (short anchors instead of full old_text). To relocate content between headings..." which flows naturally. The add-before-delete guidance is intact.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/mcp-core/tools/vault-crud-tools.ts [hunk 3]
```diff
@@ -720,6 +720,187 @@ Returns: Confirmation with lines removed and a truncated preview of the deleted
...
+Parameters:
+- start_anchor + end_anchor define a line range, not a text range — each anchor locates a full line, and entire lines are replaced (never cuts mid-line). Omit end_anchor for a single-line replace.
+- end_anchor is searched at or after the start line, so the span can never run backward. If both match the same line, only that one line is replaced.
+- content replaces the entire matched span. Must be non-empty — use vault_delete_span to delete without replacement.
+- first_match applies to both anchors independently — when an anchor matches multiple lines, takes the first instead of erroring. Blank-line runs left by the replacement are collapsed automatically.
+
+Errors:
+- "note not found" — verify path with vault_list_notes
+- "anchor not found" — fragment not on any line; verify with vault_read_note
+- "ambiguous start anchor …" / "ambiguous end anchor …" — the anchor matches multiple lines; use a longer fragment or set first_match: true
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/mcp-core/tools/vault-crud-tools.ts:776]: This is intentional — the non-empty constraint is already documented in the Parameters section: "Must be non-empty — use vault_delete_span to delete without replacement." The `.min(1)` Zod schema enforces this at the protocol level before the handler runs, making the data-layer check unreachable via MCP. Listing an unreachable error in the Errors section would be misleading.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/vault-operations/vault-patcher.ts [hunk 4]
```diff
@@ -489,9 +489,157 @@ const deleteSpan = async (
...
+      anchor: truncateForMessage(anchor),
+      position,
+      insertedLines: contentLines.length,
+      beforeBytes,
+      afterBytes,
+    })
+    const lineWord = contentLines.length === 1 ? "line" : "lines"
+    return `Inserted ${contentLines.length} ${lineWord} ${position} anchor in ${path}`
+  })
+}
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/vault-patcher.ts:638]: Valid — added a test that inserts content with 3+ consecutive internal blank lines (`"para one\n\n\n\npara two"`) and verifies they are preserved verbatim. Commit `1d8a40a`.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/mcp-core/tools/vault-crud-tools.ts [hunk 5]
```diff
@@ -720,6 +720,187 @@ Returns: Confirmation with lines removed and a truncated preview of the deleted
...
+
+Parameters:
+- anchor locates a full line — the content is inserted as whole lines before or after it (never splits a line).
+- position: "before" inserts above the anchor line; "after" inserts below it.
+- first_match: when the anchor matches multiple lines, takes the first instead of erroring.
+
+Errors:
+- "note not found" — verify path with vault_list_notes
+- "anchor not found" — fragment not on any line; verify with vault_read_note
+- "ambiguous start anchor …" — the anchor matches multiple lines; use a longer fragment or set first_match: true
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/mcp-core/tools/vault-crud-tools.ts]: Already fixed — Phase 2 (code quality) made `resolveAnchorLine`'s `role` parameter optional. `insertAtAnchor` omits it, producing "anchor not found" / "ambiguous anchor" without the misleading "start" qualifier. The tool description's Errors section was updated in the same commit.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/vault-operations/vault-patcher.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/vault-patcher.ts:560]: Duplicate of thread 1 — `collapseBlankRuns` on the full body is a deliberate design decision. See the reply on that thread.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/vault-operations/vault-patcher.ts [hunk 6]
```diff
@@ -489,9 +489,157 @@ const deleteSpan = async (
...
+    )
+
+    const anchorLine = resolveAnchorLine({
+      lines,
+      anchor,
+      fromLine: 0,
+      firstMatch,
+      path,
+      role: "start",
+    })
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/vault-patcher.ts:619]: Duplicate of thread 5 — already fixed by Phase 2. `insertAtAnchor` omits the `role` parameter, producing generic "anchor" labels in errors.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

**Comment by aliasunder** [comment]
**Re: "Add positive-plus-missing anchor test for insertAtAnchor"** — `resolveAnchorLine` performs simple `line.includes(anchor)` substring matching. There is no sentence-boundary math, state machine, or resolver logic that would produce the described failure scenario. The finding appears to describe complexity that doesn't exist in the implementation.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

**Comment by aliasunder** [comment]
**Re: "Restore removed module-layering paragraph in AGENTS.md"** — False positive. The PR's only change to AGENTS.md is a one-line update to the structure tree comment (`9 tools` → `11 tools`). No paragraph was removed from the Module layering section — `git diff main -- AGENTS.md` confirms a single-line diff. The module-layering conventions are intact.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

**Comment by aliasunder** [comment]
@coderabbit review

**Comment by aliasunder** [comment]
**Re: "Update vault-patcher module header to mention anchor operations"** — Fixed in 6e2a2d4. The header now reads "Surgical note editing — heading-targeted patches, find-and-replace, and anchor-targeted line spans (delete, replace, insert)", naming all three anchor operations the module owns.

---
*🔍 ship-check · pr-monitor · claude-fable-5*

File: src/vault-mcp/mcp-core/tools/vault-crud-tools.ts [hunk 7]
```diff
@@ -750,6 +751,195 @@ Returns: Confirmation with lines removed and a truncated preview of the deleted
...
+
+Example: vault_insert_at_anchor({ path: "Tracker.md", anchor: "| 2024-03-02 | Acme", position: "after", content: "| 2024-03-03 | Beta Corp | New entry |" }) — inserts a new table row after the matched row.
+Example: vault_insert_at_anchor({ path: "Notes/Plan.md", anchor: "## Phase 2", position: "before", content: "> [!note] Phase 1 must close before this starts.\\n" }) — inserts a callout and a blank line above the Phase 2 heading.
+
+When to use: Adding content at a precise location identified by a nearby line's text, without needing to know the heading structure. Good for inserting rows into tables, adding items into lists at a specific position, or placing content relative to a known landmark line.
+Prefer vault_patch_note for heading-targeted inserts (append/prepend to a section).${whenToolEnabledText("vault_replace_span", " Prefer vault_replace_span when replacing a block rather than inserting next to it.")}
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/mcp-core/tools/vault-crud-tools.ts]: Fixed in 0902fa0 — both mentions now go through `whenToolEnabledText` (`vault_replace_in_note` in `vault_replace_span`, `vault_patch_note` in `vault_insert_at_anchor`). Same pass gated the two ungated refs in `vault_delete_span`'s description (`vault_replace_in_note`, `vault_patch_note`) so the PR's "every sibling mention in the three anchor descriptions is gated" claim actually holds.

---
*🔍 ship-check · pr-monitor · claude-fable-5*

File: src/__tests__/integration/server-integration.test.ts [hunk 8]
```diff
@@ -62,7 +62,7 @@ describe("default config", () => {
   })
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/__tests__/integration/server-integration.test.ts:62]: Fixed in 833b627 — the four section-divider comments now carry the counts the assertions below them expect (33 / 30 / 28 / 31).

---
*🔍 ship-check · pr-monitor · claude-fable-5*

File: src/vault-mcp/mcp-core/tools/vault-crud-tools.ts [hunk 9]
```diff
@@ -750,6 +751,195 @@ Returns: Confirmation with lines removed and a truncated preview of the deleted
     },
   )
 
+  registerTool(
+    TOOL_NAMES.VAULT_REPLACE_SPAN,
+    {
+      title: "Replace Span",
+      description: `Replace a contiguous block of whole lines in a note's body with new content, identified by short anchor substrings instead of the block's full text.${whenToolEnabledText("vault_delete_span", " Same anchor semantics as vault_delete_span.")} Case-sensitive matching. Properties are preserved; operates on the body only.
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/mcp-core/tools/vault-crud-tools.ts]: Fixed in 833b627 — all three anchor tools (`vault_delete_span` included, since it shares `writePatchedNote`) now state "YAML formatting may be normalized to block style on first edit" alongside the properties-preserved claim, matching `vault_patch_note` and `vault_replace_in_note`.

---
*🔍 ship-check · pr-monitor · claude-fable-5*

File: src/vault-mcp/vault-operations/__tests__/vault-patcher.test.ts [hunk 10]
```diff
@@ -3281,6 +3287,934 @@ describe("concurrent writes (exclusive lock)", () => {
   })
 })
 
+describe("replaceSpan", () => {
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/__tests__/vault-patcher.test.ts:3306]: Fixed in 833b627 — one test per span op (`deleteSpan`, `replaceSpan`, `insertAtAnchor`) seeds a note whose only occurrence of the anchor is a frontmatter value and asserts the exact not-found message.

---
*🔍 ship-check · pr-monitor · claude-fable-5*

File: src/vault-mcp/vault-operations/vault-patcher.ts [hunk 11]
```diff
@@ -489,9 +508,140 @@ const deleteSpan = async (
...
+  if (startAnchor.length === 0) {
+    throw new Error("startAnchor cannot be empty")
+  }
+  if (endAnchor !== undefined && endAnchor.length === 0) {
+    throw new Error("endAnchor cannot be empty")
+  }
+  if (content.length === 0) {
+    throw new Error("content cannot be empty")
+  }
+  assertNoControlCharacters(content, "content")
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/vault-patcher.ts:536]: Fixed in 833b627 — `replaceSpan` and `insertAtAnchor` each get the control-character test with the exact message. Mutation-checked: commenting out either `assertNoControlCharacters` call fails exactly that tool's new test.

---
*🔍 ship-check · pr-monitor · claude-fable-5*

**Comment by aliasunder** [comment]
**Re: "Pin Obsidian-syntax guidance for the two new write tools"** — Fixed in ae46147. `vault_replace_span` and `vault_insert_at_anchor` are now in `WRITE_TOOLS`, so the `it.each` Obsidian-syntax check pins both descriptions alongside the other writers (the test file runs two more cases).

---
*🔍 ship-check · pr-monitor · claude-fable-5*

File: src/vault-mcp/mcp-core/tools/vault-crud-tools.ts [hunk 12]
```diff
@@ -750,6 +751,195 @@ Returns: Confirmation with lines removed and a truncated preview of the deleted
...
+- content is inserted verbatim — blank lines inside it are kept, and a trailing newline adds a blank line after the inserted block.
+- first_match: when the anchor matches multiple lines, takes the first instead of erroring.
+
+Errors:
+- "note not found" — verify path with vault_list_notes
+- "anchor not found" — fragment not on any line; verify with vault_read_note
+- "ambiguous anchor …" — the anchor matches multiple lines; use a longer fragment or set first_match: true
+- "hidden path blocked" — the path targets a hidden (dot-prefixed) file or folder like ".obsidian/"; hidden paths are not editable, matching Obsidian
+- "concurrent write in progress" — another write to this note is in flight; re-read the note and retry
+- "content contains a control character" — content includes a non-printable control byte; remove it before writing
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/mcp-core/tools/vault-crud-tools.ts:874]: This is intentional — the Errors section lists what the tool can return to a client, and an empty `content` never reaches the handler: the input schema is `.min(1)`, so the SDK rejects it as a protocol-level invalid-params error (`-32602`) before `insertAtAnchor` runs. The data-layer guard is defence in depth for direct callers (the unit tests), not a tool failure mode, which is why `vault_replace_span` and `vault_delete_span` omit their equivalent guards too. The non-empty requirement is stated where a client sees it: the `content` parameter description.

---
*🔍 ship-check · pr-monitor · claude-fable-5*
